OBSDATA-483: Adapt OpenCensus and OpenTelemetry extensions to the introduction of SettableByteEntity

kkonstantine opened the pull request:
<!-- Thanks for trying to help us make Apache Druid be the best it can be! Please fill out as much of the following information as is possible (where relevant, and remove it when irrelevant) to help make the intention and scope of this PR clear in order to ease review. -->

<!-- Please read the doc for contribution (https://github.com/apache/druid/blob/master/CONTRIBUTING.md) before making this PR. Also, once you open a PR, please _avoid using force pushes and rebasing_ since these make it difficult for reviewers to see what you've changed in response to their reviews. See [the 'If your pull request shows conflicts with master' section](https://github.com/apache/druid/blob/master/CONTRIBUTING.md#if-your-pull-request-shows-conflicts-with-master) for more details. -->

Fixes [OBSDATA-483](https://confluentinc.atlassian.net/browse/OBSDATA-483) 


<!-- If you are a committer, follow the PR action item checklist for committers:
https://github.com/apache/druid/blob/master/dev/committer-instructions.md#pr-and-issue-action-item-checklist-for-committers. -->

### Description

In our deployments, metrics emitted from a Kafka topic can be a mix of formats between OpenTelemetry and OpenCensus formats. Up until version 24.0.0, the entity passed as input to the format readers could be cast as `ByteEntity` and have the protobuf message parsed directly from the `ByteBuffer` of this entity. However with druid-24.0.0 a wrapper class `SettableByteEntity` is introduced and now the classes in the opencensus and opentelemtry extensions need to be adapted to handle `SettableByteEntity` objects that use composition instead of inheritance in relation to `ByteEntity` objects. 

<!-- Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's not necessary to repeat the description here, however, you may choose to keep one summary sentence. -->

<!-- Describe your patch: what did you change in code? How did you fix the problem? -->

<!-- If there are several relatively logically separate changes in this PR, create a mini-section for each of them. For example: -->

<!--
In each section, please describe design decisions made, including:
 - Choice of algorithms
 - Behavioral aspects. What configuration values are acceptable? How are corner cases and error conditions handled, such as when there are insufficient resources?
 - Class organization and design (how the logic is split between classes, inheritance, composition, design patterns)
 - Method organization and design (how the logic is split between methods, parameters and return types)
 - Naming (class, method, API, configuration, HTTP endpoint, names of emitted metrics)
-->


<!-- It's good to describe an alternative design (or mention an alternative name) for every design (or naming) decision point and compare the alternatives with the designs that you've implemented (or the names you've chosen) to highlight the advantages of the chosen designs and names. -->

<!-- If there was a discussion of the design of the feature implemented in this PR elsewhere (e. g. a "Proposal" issue, any other issue, or a thread in the development mailing list), link to that discussion from this PR description and explain what have changed in your final design compared to your original proposal or the consensus version in the end of the discussion. If something hasn't changed since the original discussion, you can omit a detailed discussion of those aspects of the design here, perhaps apart from brief mentioning for the sake of readability of this PR description. -->

<!-- Some of the aspects mentioned above may be omitted for simple and small changes. -->

<hr>

##### Key changed/added classes in this PR

* Since we need to choose during runtime the appropriate reader, a new class `HybridProtobufReader` is introduced that can delegate reading and parsing of the input entity to appropriate reader (OpenCensus or OpenTelemetry)
* Given that samplers from the UI still pass a `KafkaRecordEntity` directly and not a `SettableByteEntity` the code is adapted for now to distinguish between the two inputs by checking the instance of the input (a potential near future fix is to adapt samplers to also use `SettableByteEntity`)
* Advancement of the position of the entity buffer is done explicitly since the protobuf parser does not manipulate the position of the `ByteBuffer` that is passed to it as an arguemnt. 
* Logging of messages is enabled in tests for opencensus
* Adaptation of current tests seems sufficient at this point. Eventually it'd be good to add integration tests that ingest protobuf messages from Kafka with a mix of OpenCensus and OpenTelemetry formats. 

<hr>

<!-- Check the items by putting "x" in the brackets for the done things. Not all of these items apply to every PR. Remove the items which are not done or not relevant to the PR. None of the items from the checklist below are strictly necessary, but it would be very helpful if you at least self-review the PR. -->

This PR has:
- [x] been self-reviewed.
   - [x] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.


[OBSDATA-483]: https://confluentinc.atlassian.net/browse/OBSDATA-483?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[OBSDATA-483]: https://confluentinc.atlassian.net/browse/OBSDATA-483?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

kkonstantine commented:
Tested by adapting existing tests. Will check test coverage to see if we are still missing any paths in the relevant classes.

kkonstantine commented:
A few more tests might be granted but taking this to a full review mode given that the changes are already being tested in staging.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputFormat.java:
```diff
@@ -73,41 +74,22 @@ public boolean isSplittable()
-    catch (Throwable t) {
-      // assume input is opencensus if something went wrong
+    else {
+      SettableByteEntity<ByteEntity> wrapper = new SettableByteEntity<>();
```
nit, you can avoid the intermediate variable, and set `settableEntity` directly here

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java:
```diff
@@ -81,29 +97,75 @@ public OpenCensusProtobufReader(
   {
     try {
-      return parseMetric(Metric.parseFrom(source.getBuffer()));
+      return parseMetric(Metric.parseFrom(source.open()));
```
since we have a ByteEntity, we can avoid the overhead of parsing the underlying ByteBuffer via InputStream and read it directly like we did before using
```suggestion
      return parseMetric(Metric.parseFrom(source.getEntity().getBuffer()));
```

xvrl replied:
looking a bit at the protobuf code, it looks like it has some pretty sophisticated optimizations for ByteBuffers and will read the underlying array directly or bypass the JVM bounds check using unsafe if possible to get the best performance.
It is likely that this causes the ByteBuffer position to not advance and trigger the check here https://github.com/apache/druid/blob/6346b9561df7a92557acb51a56585084a0eb8633/indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SettableByteEntity.java#L97-L99

A simple workaround might be to call `buffer.position(buffer.position() + buffer.remaining())` when done parsing from the buffer.

While the additional check is nice, I think it breaks the existing "contract" in that previously there was no requirement to read out the buffer entirely,. I believe it was mainly added to avoid some kind of unexpected bug or concurrency issue with SettableByteEntity

kkonstantine replied:
I don't think we have performance issues by calling `open` here. If we look at the code of [SettableByteEntity#open](https://github.com/confluentinc/druid/blob/e0c3831aed8f8b904719d8c14b18284e5c72f3f9/indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SettableByteEntity.java#LL80C9-L80C9) all it does is to check if the input stream is already open and if it doesn't it returns it. (I want to say I checked that before I write the call here but it's been a few days).

Looking there at [SettableByteEntity#setEntity](https://github.com/confluentinc/druid/blob/major-druid-upgrade-confluent-rc2/indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SettableByteEntity.java#L56-L61) which I assume is called every time before a call to `open` in our case, it seems that it does something similar to what you describe. It sets the buffer for that inputstream and resets the `opened` flag. So, in a sense it re-uses the bytebuffer that is given via the `entity` that is passed. 

So not a traditional `open` with its costs it seems (but I still might be missing something given my low familiarity with this code). Wdyt?

xvrl replied:
I think it's still more efficient for the protobuf decoder to have access to the buffer, since it can skip ahead rather than having to consume all the bytes in a linear fashion from the inputstream. The InputStream will also incur the cost of bounds checks on the ByteBuffer, which the protobuf decoder currently goes out of its way to avoid.

xvrl replied:
the fact that SettableByteEntity seems geared towards returning an InputStream is something we'll probably want to improve upstream. Part of the reason we have ByteEntity is so that we can optimize access to the underlying buffer for performance. I don't think anything stands in our way here to get the underlying ByteBuffer now, and we can improve the API to make that easier in the future.

kkonstantine replied:
I see what you mean now. I didn't go into that depth with the protobuf calls (still a bit challenging to get to the non-sequential parsing you are referring to). 

The main other example with protobuf beyond the classes we discuss here that I can find is in 
[ProtobufReader#intermediateRowIterator](https://github.com/confluentinc/druid/blob/e0c3831aed8f8b904719d8c14b18284e5c72f3f9/extensions-core/protobuf-extensions/src/main/java/org/apache/druid/data/input/protobuf/ProtobufReader.java#LL83)
and while it passes a ByteBuffer it first reads and converts the stream into a bytearray. 

```
  Iterators.singletonIterator(protobufBytesDecoder.parse(ByteBuffer.wrap(IOUtils.toByteArray(source.open()))))
```

kkonstantine replied:
Changed the code to set the buffer position explicitly to the limit. Let me know if the code now matches what you had in mind.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java:
```diff
@@ -81,29 +97,75 @@ public OpenCensusProtobufReader(
+      return parseMetric(Metric.parseFrom(source.open()));
     }
-    catch (InvalidProtocolBufferException e) {
+    catch (IOException e) {
```
using the suggestion above we can also avoid having to catch additional IOExceptions

xvrl replied:
looks like we revert this change now and only catch InvalidProtocolBufferException?

kkonstantine replied:
I reinstated `InvalidProtocolBufferException` but that doesn't remove the IOException in `sample` and `read` methods.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java:
```diff
@@ -217,7 +279,7 @@ private void addPointRows(Point point, Metric metric, LabelContext labelContext)
 
   @Override
-  public CloseableIterator<InputRowListPlusRawValues> sample()
+  public CloseableIterator<InputRowListPlusRawValues> sample() throws IOException
```
this change is unnecessary if we read the ByteBuffer directly as suggested above

xvrl replied:
looks like we can avoid throwing checked exceptions here now

kkonstantine replied:
This is thrown by the `read` method in the base interface. Doesn't have to do anything with the use of `open` after all. If we want to use try-with-resources which is good IMO, we need to add this exception to the method declaration. 

This is declared in: 
https://github.com/confluentinc/druid/blob/94001257689b8e423c438e6a9e2f4197529fc6aa/extensions-core/kafka-indexing-service/src/main/java/org/apache/druid/data/input/kafkainput/KafkaInputReader.java#L143

while elsewhere the overriding method omits the exception in the signature. 

The interface is: https://github.com/confluentinc/druid/blob/0.22.1-confluent/core/src/main/java/org/apache/druid/data/input/InputEntityReader.java
and hasn't changed in recent versions.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java:
```diff
@@ -81,29 +97,75 @@ public OpenCensusProtobufReader(
 
   @Override
-  public CloseableIterator<InputRow> read()
+  public CloseableIterator<InputRow> read() throws IOException
```
this change is unnecessary if we read the ByteBuffer directly as suggested below

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputFormat.java:
```diff
@@ -73,41 +74,22 @@ public boolean isSplittable()
-          source.getClass().getClassLoader(),
-          OpenCensusProtobufInputFormat.VERSION_HEADER_KEY
-      );
+    // Sampler passes a KafkaRecordEntity directly, while the normal code path wraps the same entity in a
+    // SettableByteEntity
```
it might be worth making a change upstream to have the sampler also pass in a SettableByteEntity to avoid having to handle this edge.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputRowParser.java:
```diff
@@ -103,10 +104,13 @@ public OpenCensusProtobufInputRowParser withParseSpec(ParseSpec parseSpec)
-        new ByteEntity(input),
+        settableByteEntity,
         metricDimension,
+        null,
```
it's bit odd to pass null here, since this would likely break if the downstream code ever tried to use this to write a column with a null name.

It looks like there is actually a bug in the existing OpenCensusProtobufInputFormat here since it does not actually ever use the valueDimension field when decoding OpenCensus and instead hard-codes it as `value`

I think we can fix that and have OpenCensusProtobufReader honor the `valueDimension` setting so we don't have to pass `null` here. This should not break any backwards compatibility, since we do not set the `valueDimension` anywhere in our configs today, and the default for valueDimension is `value`

kkonstantine replied:
https://confluentinc.atlassian.net/browse/OBSDATA-566

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java:
```diff
@@ -81,29 +97,75 @@ public OpenCensusProtobufReader(
+
+  public ProtobufReader whichReader()
+  {
+    // assume InputEntity is always defined in a single classloader (the kafka-indexing-service classloader)
+    // so we only have to look it up once. To be completely correct we should cache the method based on classloader
+    MethodHandle getHeaderMethod = KafkaUtils.lookupGetHeaderMethod(
+        source.getEntity().getClass().getClassLoader(),
+        VERSION_HEADER_KEY
+    );
```
as a follow-up improvement, we might want to cache the `getHeaderMethod` so we don't have to do the lookup on every record. Once we have successfully looked up the method handle once we shouldn't have to do it again, since that's probably not a cheap call to make.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java:
```diff
@@ -81,29 +97,75 @@ public OpenCensusProtobufReader(
+        );
+      case OPEN_CENSUS:
+      default:
+        return new OpenCensusProtobufReaderInternal();
```
rather than creating an internal reader for this particular case, what would you think of keeping this code path more symmetrical by moving the code that handles the format switching to a new class (e.g. `HybridProtobufReader`), and then have that call either `new OpenTelemetryMetricsProtobufReader(...)` or `new OpenCensusProtobufReader(...)` so we don't pollute the OpenCensus reader code with otel-specific handling, and keep the Otel and OC readers clean counterparts of each-other.

xvrl commented on extensions-contrib/opentelemetry-extensions/src/main/java/org/apache/druid/data/input/opentelemetry/protobuf/OpenTelemetryMetricsProtobufReader.java:
```diff
@@ -80,27 +78,15 @@ public OpenTelemetryMetricsProtobufReader(
   @Override
   public CloseableIterator<InputRow> read()
   {
-    Supplier<Iterator<InputRow>> supplier = Suppliers.memoize(() -> readAsList().iterator());
```
it looks like we might want to keep the memoizing supplier here, since it was added to avoid the scenario where exceptions are thrown before we actually call next(), see https://github.com/confluentinc/druid/pull/99

As you had pointed out in that PR we should have added a test so that we wouldn't accidentally regress here.

xvrl commented on extensions-contrib/opentelemetry-extensions/src/main/java/org/apache/druid/data/input/opentelemetry/protobuf/OpenTelemetryMetricsProtobufInputFormat.java:
```diff
@@ -63,9 +64,20 @@ public boolean isSplittable()
+      settableEntity = (SettableByteEntity<? extends ByteEntity>) source;
+    }
+    else {
+      SettableByteEntity<ByteEntity> wrapper = new SettableByteEntity<>();
```
nit, you can avoid the intermediate variable, and set settableEntity directly here

xvrl commented on extensions-contrib/opentelemetry-extensions/src/main/java/org/apache/druid/data/input/opentelemetry/protobuf/OpenTelemetryMetricsProtobufReader.java:
```diff
@@ -80,27 +78,15 @@ public OpenTelemetryMetricsProtobufReader(
   {
     try {
-      return parseMetricsData(MetricsData.parseFrom(source.getBuffer()));
+      return parseMetricsData(MetricsData.parseFrom(source.open()));
```
same comment about using the buffer instead of the input stream

kkonstantine reviewed (commented):
Thanks for the review @xvrl !
I believe I didn't miss anything with a change or an answer. Will have to deploy on staging too to confirm some of the changes that were rolled back. Didn't revert the call to `open` because I'm pretty sure this will break things.

kkonstantine commented:
Just rebased on top of 24.0.1

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufInputFormat.java:
```diff
@@ -49,8 +45,6 @@ public class OpenCensusProtobufInputFormat implements InputFormat
   private final String metricLabelPrefix;
```
VERSION_HEADER_KEY and OPENTELEMETRY_FORMAT_VERSION constants above are no longer used and can be removed

kkonstantine replied:
I thought I had removed those. Good catch. Ran the inspection too

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/HybridProtobufReader.java:
```diff
@@ -0,0 +1,141 @@
+        int version =
+            ByteBuffer.wrap(versionHeader).order(ByteOrder.LITTLE_ENDIAN).getInt();
+        if (version == OPENTELEMETRY_FORMAT_VERSION) {
+          return OPEN_TELEMETRY;
```
nit, for consistency we might want to keep OPENTELEMETRY as one word in constants (same with OPENCENSUS)

kkonstantine replied:
Makes total sense.

kkonstantine commented:
Rebased on top of `druid-24.0.2` that now has been released (by rebasing the underlying branch) and reverted the changes of the tests that should not expect a call to `open()` anymore.

xvrl commented on extensions-contrib/opentelemetry-extensions/src/main/java/org/apache/druid/data/input/opentelemetry/protobuf/OpenTelemetryMetricsProtobufReader.java:
```diff
@@ -221,8 +228,10 @@ InputRow createRow(long timeUnixMilli, Map<String, Object> event)
 
   @Override
-  public CloseableIterator<InputRowListPlusRawValues> sample()
+  public CloseableIterator<InputRowListPlusRawValues> sample() throws IOException
```
same here we can avoid throwing IOException here now

kkonstantine replied:
replied above. It's required by `try-with-resources`

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/HybridProtobufReader.java:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  @Override
+  public CloseableIterator<InputRowListPlusRawValues> sample() throws IOException
```
same here we can avoid throwing IOException here now

kkonstantine reviewed (commented):
Thanks for the follow up @xvrl 
I added back `InvalidProtocolBufferException` but left a comment on why `IOException` can't be removed without an error if we are to keep `try-with-resources`.

kkonstantine commented:
Happy to follow up on outstanding comments. 
In the meantime I have updated the description. 
We also got a green build with tests passing 🎉

xvrl commented on .semaphore/semaphore.yml:
```diff
@@ -4,6 +4,10 @@ agent:
   machine:
     type: s1-prod-ubuntu20-04-amd64-1
 
+# Increase time limit to allow tests to run to completion
+execution_time_limit:
+  hours: 2
```
I'm surprised it takes that long now, we used to be able to run tests much faster than that in the past, but we can leave that until we figure out what it taking so long

kkonstantine replied:
A similar setting on another repo has a comment: `# Increase time limit to occasionally allow for cache population.` 
Which could have been the case here as well, given that the successful run took 40min. Reverting this patch and we can reconsider if this starts happening more frequently.

xvrl commented on extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java:
```diff
@@ -217,8 +225,10 @@ private void addPointRows(Point point, Metric metric, LabelContext labelContext)
-    return read().map(row -> InputRowListPlusRawValues.of(row, ((MapBasedInputRow) row).getEvent()));
+    try (CloseableIterator<InputRow> iterator = read()) {
+      return iterator.map(row -> InputRowListPlusRawValues.of(row, ((MapBasedInputRow) row).getEvent()));
+    }
```
are we sure we should use try-with-resources here? If we close the underlying iterator before returning the mapped one, wouldn't this cause issues? It seems it should be the responsibility of the called to close the returned iterator, not this method.

kkonstantine replied:
You're right. Only that [InputEntityIteratingReader](https://github.com/confluentinc/druid/blob/0.22.1-confluent/core/src/main/java/org/apache/druid/data/input/impl/InputEntityIteratingReader.java#L87-L99) which is one of the few users of this call doesn't close it as it should when it handles the `IOException`. 

I guess it doesn't blow up because of low coverage of sample methods (I found out manually on the UI). Reverting to unblock this for now and we can reconsider in the future.

kkonstantine replied:
Btw, in the Hybrid class, IOException has to stay because we are calling the interface method (that has it)

xvrl replied:
good catch about [InputEntityIteratingReader](https://github.com/confluentinc/druid/blob/0.22.1-confluent/core/src/main/java/org/apache/druid/data/input/impl/InputEntityIteratingReader.java#L87-L99) maybe you can file an issue upstream to raise that potential problem
